### DISCUSSION TITLE
v2 - Is it possible to set Livewire's updated() hook when using Resource Forms?

### DISCUSSION THREAD

**Discussion by emcarson** [filamentphp/filament#7212] [Help] [answered]
Hi everybody,

I'm working on a form that has many computed fields (calculations).
Each of these calculated fields depends on multiple other fields, which in turn may be calculations themselves.

For example (simplified for clarity):

- total_cost: depends on shipping_method, round_weight and weight
- price: depends on shipping_method, non_commissionable_adjustment, commissionable_adjustment, fee
- gross_income: depends on price, total_cost

The approaches I tried so far:

**1) Using `reactive()`, `afterStateUpdated()` and `$set()` on the parent fields.**

It quickly becomes super confusing since I have to repeat the same formulas on multiple fields (ie: I need to calculate total_cost and set it from shipping_method, round_weight and weight).

I have no way of updating grandchildren fields automatically (now shipping_method must be responsible of calculating and setting price and gross_income).

**2) Using a `Placeholder` field.**

For me it's not and option since I need to persist the data and allow the user to customize some of the calculated fields manually before storing.

**3) Using Livewire's `updated()` hook.**

This would be an ideal solution since I can recalculate all fields automatically whenever a reactive field changes.
Problem is, I can't find a way to access it when using the Resource's Form Builder.

I could create a Custom Livewire component but after reading the docs I'm not sure how to register it / replace the default Resource actions with it.

Sorry for the long post, thank you for hearing me out.

**Comment by leandrocfe** [comment, marked as answer]
Regarding the 3, you can use the updated method in the CreatePage.php and EditPage.php files.
I'm not sure if it's the right way to do that, but you can set the fields to `->lazy()` or `->live(onBlur: true)` and use this.

```php
public function updated($name, $value): void
{
    dd($name, $value);
}
```

**Reply by emcarson** [reply]
Amazing, it works great, thank you @leandrocfe!

I ended up creating a Trait to share functionality between the Create and Edit pages:

```php
<?php

namespace App\Filament\Resources\OrderResource\Pages;

use App\Concerns\CalculatesOrder;
use App\Filament\Resources\OrderResource;
use Filament\Resources\Pages\CreateRecord;

class CreateOrder extends CreateRecord
{
    use CalculatesOrder;

    protected static string $resource = OrderResource::class;

    public function updated($name, $value): void
    {
        if ($name === 'data.shipping_method_id') {
            $this->data['shipment_id'] = '';
        }

        $this->calculate();
    }

    protected function beforeValidate(): void
    {
        $this->calculate();
    }
}
```

```php
<?php

namespace App\Concerns;

use App\Models\ShippingMethod;
use JetBrains\PhpStorm\NoReturn;

/**
 * @property \App\Models\Order $record
 * @property array $data
 */
trait CalculatesOrder
{
    #[NoReturn]
    protected function calculate(): void
    {
        $shippingMethod = $this->record->shippingMethod ?? ShippingMethod::find($this->data['shipping_method_id']);

        $weight = $this->getWeight();
        $totalCost = $this->getTotalCost($shippingMethod, $weight);

        $this->data = array_merge($this->data, [
            'total_cost' => $totalCost,
        ]);
    }
}
```

**Reply by emcarson** [reply]
PS: I had to move the `afterStatusUpdated()` logic from the Form definition to the `updated()` method because I needed specific fields to update before the calculations.  I'm also calling `calculate()` before validation, just to ensure that the values are always recalculated (even if the user submits the edit form without changing anything).
